### PR TITLE
[PDS-205737] Invalidation Train

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Backup and restore workflows need to account for the possibility that, during a restore, a running service will read data that has been written before the clean-transactions-range operation has run (which is uncommitted). Typically this is done by locking down KVS access. However, this is clearly insufficient where lock watches may be used.
- Note that this currently isn't an issue just yet because validation of the Atlas cache is at 100 percent, but it's best we have this in place now, rather than find ourselves in corruption land later on.

**Implementation Description (bullets)**:
- Add a new endpoint that removes service implementations from timelock's tracking map. Future user requests will re-create these, but they have a different leadership ID, different lock watch ID, and all locks will be invalidated through this.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Added fairly standard tests for the new endpoint and all associated functionality.

**Concerns (what feedback would you like?)**:
- Is there some bizarre race condition I might have missed?
- Is there a resource leak I might have missed?

**Where should we start reviewing?**: TimelockNamespaces

**Priority (whenever / two weeks / yesterday)**: this week please

@leonz / @j-baker for SA

### CODE REVIEW

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/TimelockNamespacesTest.java [hunk 1]
```diff
@@ -154,6 +155,29 @@ public void pathsForTimelockAndLockWatchServicesAreNotValid() {
...
+                .thenReturn(mock(TimeLockServices.class));
+
+        String client = uniqueClient();
+        TimeLockServices services = namespaces.get(client);
+        namespaces.invalidateResourcesForClient(client);
+        verify(services).close();
+
+        services = namespaces.get(client);
+        namespaces.invalidateResourcesForClient(client);
+        verify(services).close();
```
Comment (sudiksha27) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/TimelockNamespacesTest.java:172]: can we add an assertion that 2 service objects were different?

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/TimelockNamespacesTest.java [hunk 2]
```diff
@@ -154,6 +155,29 @@ public void pathsForTimelockAndLockWatchServicesAreNotValid() {
...
+    }
+
+    @Test
+    public void handlesInvalidationOfNonexistentClients() {
+        assertThatCode(() -> namespaces.invalidateResourcesForClient("somethingUnknown"))
+                .doesNotThrowAnyException();
```
Comment (sudiksha27) [inline timelock-impl/src/test/java/com/palantir/atlasdb/timelock/TimelockNamespacesTest.java:178]: 💯

**Review by sudiksha27 — APPROVED** [review]
Solid PR - I like the final solution! Tests are good - cannot think of a race condition we may be missing so LGTM!

File: timelock-impl/src/test/java/com/palantir/atlasdb/timelock/TimelockNamespacesTest.java (hunk 1 above)
Comment (jeremyk-91) [reply timelock-impl/src/test/java/com/palantir/atlasdb/timelock/TimelockNamespacesTest.java:172]: Yep, will do.

**Pull request merged**